data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/FormatDateTimeExpressionFunction.java

```diff
@@ -0,0 +1,103 @@
+ * compatible open source license.
+ *
+ */
+
+package org.opensearch.dataprepper.expression;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAccessor;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.inject.Named;
+
+import org.opensearch.dataprepper.model.event.Event;
+
+@Named
+public class FormatDateTimeExpressionFunction implements ExpressionFunction {
+
+    @Override
+    public String getFunctionName() {
+        return "formatDateTime";
+    }
+
+    @Override
+    public Object evaluate(List<Object> args, Event event, Function<Object, Object> convertLiteralType) {
+        if (args.size() < 2) {
+            throw new IllegalArgumentException(getFunctionName() + "() takes at least 2 arguments");
+        }
+        if (args.size() > 4) {
+            throw new IllegalArgumentException(getFunctionName() + "() takes at most 4 arguments");
+        }
+        List<String> argStrings;
+        try {
+            argStrings = args.stream()
+                    .map(String.class::cast)
+                    .map(String::trim)
+                    .collect(Collectors.toUnmodifiableList());
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Arguments in " + getFunctionName() + "() function should be of Json Pointer type or String type");
+        }
+
+        String eventKey = argStrings.get(0);
+        String pattern = argStrings.get(1);
+        pattern = unquote(pattern);
+
+
+        ZoneId destinationTimeZone = ZoneOffset.UTC;
+        if (argStrings.size() > 2) {
+            String destinationZoneIdArg = unquote(argStrings.get(2));
+            try {
+                destinationTimeZone = ZoneId.of(destinationZoneIdArg);
+            } catch (Exception e) {
+                throw new IllegalArgumentException("Destination time zone [" + destinationZoneIdArg + "] is invalid");
+            }
+        }
+
+        ZoneId sourceTimeZone = ZoneOffset.UTC;
+
+        if (argStrings.size() > 3) {
+            String sourceTimeZoneArg = unquote(argStrings.get(3));
+            try {
+                sourceTimeZone = ZoneId.of(sourceTimeZoneArg);
+            } catch (Exception e) {
+                throw new IllegalArgumentException("Source time zone [" + sourceTimeZoneArg + "] is invalid");
+            }
+        }
+
+        DateTimeFormatter formatter;
+        try {
+            formatter = DateTimeFormatter.ofPattern(pattern).withZone(destinationTimeZone);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Date pattern [" + pattern + "] is invalid");
+        }
+
+        Object target = event.get(eventKey, Object.class);
+        if (target instanceof Number) {
```

Comment:
Sorry since files were removed the orginal conversation was lost.

RE: https://github.com/opensearch-project/data-prepper/pull/5982#discussion_r2320060293

> @san81 @dlvenable One question. While the code above is correct from a logical standpoint, how is that possible in practice? An `Event` stores all the values in a JSON serialized format, which for data types above means either a number or an array (see https://github.com/FasterXML/jackson-modules-java8/tree/2.x/datetime ). And unfortunately there is no way to tell which specific data type that number or array represents.
> 
> I think we should only handle numbers (to use with `Instant.ofEpochMilli`) or strings (to use with `DateTimeFormatter.ISO_DATE_TIME.parse`). Please let me know what you think.